--- a/klippy/extras/filament_feed.py
+++ b/klippy/extras/filament_feed.py
@@ -187,8 +187,10 @@
             self.white_light.set_pwm(print_time, 0)
 
 class FeedPort:
-    def __init__(self, printer, reactor, pin, threshold):
+    def __init__(self, printer, reactor, pin, threshold, index):
         self.reactor = reactor
+        self.index = index
+        self.ace = None
         ppins = printer.lookup_object('pins')
         self._port = ppins.setup_pin('adc', pin)
         self._port_adc_value = 0
@@ -226,7 +228,12 @@
     def get_adc_value(self):
         return self._port_adc_value
 
+    def add_ace(self, ace):
+        self.ace = ace
+
     def get_filament_detected(self):
+        if self.ace is not None:
+            return self.ace.gate_status[self.index] == 1
         return self._filament_detected
 
 class FeedTachometer:
@@ -412,12 +419,12 @@
         self._port = []
         tmp_pin = config.get('port_ch_1_pin')
         threshold = config.getfloat('port_ch_1_threshold')
-        tmp_obj = FeedPort(self.printer, self.reactor, tmp_pin, threshold)
+        tmp_obj = FeedPort(self.printer, self.reactor, tmp_pin, threshold, FEED_CHANNEL_1)
         tmp_obj.register_cb_2_port_event(self._port_ch1_event_handler)
         self._port.append(tmp_obj)
         tmp_pin = config.get('port_ch_2_pin')
         threshold = config.getfloat('port_ch_2_threshold')
-        tmp_obj = FeedPort(self.printer, self.reactor, tmp_pin, threshold)
+        tmp_obj = FeedPort(self.printer, self.reactor, tmp_pin, threshold, FEED_CHANNEL_2)
         tmp_obj.register_cb_2_port_event(self._port_ch2_event_handler)
         self._port.append(tmp_obj)
         self.gcode.register_mux_command("FEED_PORT", "MODULE",
@@ -542,6 +549,9 @@
     def _ready(self):
         self.toolhead = self.printer.lookup_object('toolhead')
         self.gcode_move = self.printer.lookup_object('gcode_move')
+        self.ace = self.printer.lookup_object('ace', None)
+        for i in self._port:
+            i.add_ace(self.ace)
         self.exception_manager = self.printer.lookup_object('exception_manager', None)
         self.reactor.update_timer(self._check_init_state_timer,
                                   self.reactor.monotonic() + 2 * FEED_PORT_ADC_REPORT_TIME)
@@ -562,7 +572,7 @@
         self.reactor.unregister_timer(self._check_init_state_timer)
 
         for ch in range(FEED_CHANNEL_NUMS):
-            if self._port[ch].get_adc_value() < FEED_PORT_ADC_VAL_MODULE_EXIST:
+            if self._port[ch].get_adc_value() < FEED_PORT_ADC_VAL_MODULE_EXIST or self.ace is not None:
                 self.module_exist[ch] = True
             else:
                 self.module_exist[ch] = False
@@ -994,8 +1004,33 @@
                         motor_err_max_cnt = FEED_LOAD_MOTOR_ERR_CNT_MAX
                         wheel_err_max_cnt = FEED_LOAD_WHEEL_ERR_CNT_MAX
                         one_step_cnt = self.wheel[ch].ppr * 2.0 * 10.0 / FEED_WHEEL_CIRCUMFERENCE
+                        if self.ace is not None:
+                            self.ace._feed(self.filament_ch[ch], self.ace.load_lengths[self.filament_ch[ch]], self.ace.load_speed, 0)
 
                         while 1:
+                            if self.ace is not None:
+                                self.reactor.pause(self.reactor.monotonic() + 0.105)
+                                systime_2 = self.reactor.monotonic()
+                                port_detect = self._port[ch].get_filament_detected()
+                                runout_detect = self.runout_sensor[ch].get_status(0)['filament_detected']
+                                logging.info("[feed_loading] ace load: slot:%d, port:%s, runout:%s",
+                                             self.filament_ch[ch], port_detect, runout_detect)
+                                if runout_detect == True:
+                                    self.ace._stop_feeding(self.filament_ch[ch])
+                                    self.ace.wait_ace_ready()
+                                    self.channel_error[ch] = FEED_OK
+                                    break
+                                if port_detect == False:
+                                    self.ace._stop_feeding(self.filament_ch[ch])
+                                    self.channel_error[ch] = FEED_ERR_NO_FILAMENT
+                                    self.exception_code[ch] = 33
+                                    break
+                                if systime_2 - systime_0 > FEED_LOAD_TIMEOUT_TIME:
+                                    self.ace._stop_feeding(self.filament_ch[ch])
+                                    self.channel_error[ch] = FEED_ERR_TIMEOUT
+                                    self.exception_code[ch] = 34
+                                    break
+                                continue
                             wheel_cnt_a_1 = self.wheel[ch].get_counts()
                             wheel_cnt_b_1 = self.wheel_2[ch].get_counts()
                             motor_cnt_1 = self.motor_tachometer.get_counts()
@@ -1014,6 +1049,8 @@
                                          wheel_cnt_a_0, wheel_cnt_a_1, wheel_cnt_a_2, wheel_cnt_b_0, wheel_cnt_b_1, wheel_cnt_b_2, wheel_err_max_cnt,
                                          motor_cnt_1, motor_cnt_2, motor_cnt_2 - motor_cnt_1, motor_err_max_cnt)
                             if runout_detect == True:
+                                if self.ace is not None:
+                                    self.ace._stop_feeding(self.filament_ch[ch])
                                 self.channel_error[ch] = FEED_OK
                                 break
                             if port_detect == False:
@@ -1059,6 +1096,8 @@
                         if self.channel_error[ch] != FEED_OK:
                             self._hang_neutral(ch)
                             raise ValueError('logic error!')
+                    if self.ace is not None and self.ace.uses_ace_assist():
+                        self.ace._enable_feed_assist(self.filament_ch[ch])
 
                     self.gcode.run_script_from_command("M104 S%d\r\n" % (filament_feed_temp))
                     try:
@@ -1231,6 +1270,8 @@
 
             # unload
             elif action == FEED_ACT_UNLOAD:
+                if self.ace is not None:
+                    self.ace._disable_feed_assist()
                 self.exception_code[ch] = 70
                 if stage not in [None, FEED_UNLOAD_STAGE_PREPARE, FEED_UNLOAD_STAGE_DOING,
                                  FEED_UNLOAD_STAGE_CANCEL]:
@@ -1314,6 +1355,8 @@
                         # finish
                         self.toolhead.wait_moves()
                         self.gcode.run_script_from_command("M104 S0\r\n")
+                        if self.ace is not None:
+                            self.ace.retract_fil(self.filament_ch[ch])
                         self.channel_error[ch] = FEED_OK
                         self._set_channel_state(ch, FEED_STA_UNLOAD_FINISH)
 
